lib/ClangImporter/ClangDerivedConformances.cpp

```diff
@@ -1364,3 +1364,218 @@ void swift::conformToCxxSpanIfNeeded(ClangImporter::Implementation &impl,
     impl.addSynthesizedProtocolAttrs(decl, {KnownProtocolKind::CxxMutableSpan});
   }
 }
+
+// applies to [operator op(T, T)] where T = classDecl
+static bool synthesizeCXXOperatorWithFunctionObject(
+    ClangImporter::Implementation &impl, const clang::CXXRecordDecl *classDecl,
+    const clang::CXXRecordDecl *funcDecl,
+    clang::BinaryOperatorKind operatorKind) {
+  auto &clangCtx = impl.getClangASTContext();
+  auto &clangSema = impl.getClangSema();
+
+  clang::OverloadedOperatorKind opKind =
+      clang::BinaryOperator::getOverloadedOperator(operatorKind);
+  const char *opSpelling = clang::getOperatorSpelling(opKind);
+
+  auto declName = clang::DeclarationName(&clangCtx.Idents.get(opSpelling));
+
+  // Determine the Clang decl context where the new operator function will be
+  // created. We use the translation unit as the decl context of the new
+  // operator, otherwise, the operator might get imported as a static member
+  // function of a different type (e.g. an operator declared inside of a C++
+  // namespace would get imported as a member function of a Swift enum), which
+  // would make the operator un-discoverable to Swift name lookup.
+  auto declContext =
+      const_cast<clang::CXXRecordDecl *>(classDecl)->getDeclContext();
+  while (!declContext->isTranslationUnit()) {
+    declContext = declContext->getParent();
+  }
+
+  clang::CXXMethodDecl *methodDecl = nullptr;
```

Comment:
Sorry if I'm missed this from `conformToHashableIfNeeded`, but is it ever possible for `methodDecl` to never be assigned? What guarantees the presence of a suitable `operator ()`?

Reply:
Perhaps this assumption is only valid if this helper function is called with `std::equal_to`, but I think that could/should be made clearer.

Reply:
The intention is that this function is invoked with a function object `funcDecl` (one with `operator()`). I should put an `ASSERT(methodDecl)` after the search.